File: decision-matrix/frontend/src/lib/lineSplit.ts
import type { InfraObject, InfraObjectCreate } from './api';
import { roundCoord } from './coords';
import { getLineCoordinates, isLineSubtype } from './infraGeometry';

/** Минимальное расстояние от конца линии, чтобы вставка не дублировала вершину (≈10 м). */
export const LINE_SPLIT_ENDPOINT_MIN_KM = 0.01;

/** Максимальное расстояние клика от линии для разрезания (≈300 м, как snap линий). */
export const LINE_SPLIT_HIT_TOLERANCE_KM = 0.3;

function haversineKm(lon1: number, lat1: number, lon2: number, lat2: number): number {
  const r = 6371.0;
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dp = ((lat2 - lat1) * Math.PI) / 180;
  const dl = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return r * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export type ClosestPointOnSegment = {
  point: [number, number];
  /** Параметр интерполяции 0..1 на отрезке. */
  t: number;
  distanceKm: number;
  segmentIndex: number;
};

/** Ближайшая точка на отрезке a–b к p (локальная интерполяция lon/lat). */
export function closestPointOnSegment(
  p: [number, number],
  a: [number, number],
  b: [number, number],
  segmentIndex: number,
): ClosestPointOnSegment {
  const dx = b[0] - a[0];
  const dy = b[1] - a[1];
  const len2 = dx * dx + dy * dy;
  let t = len2 > 0 ? ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / len2 : 0;
  t = Math.max(0, Math.min(1, t));
  const point: [number, number] = [a[0] + t * dx, a[1] + t * dy];
  return {
    point,
    t,
    distanceKm: haversineKm(p[0], p[1], point[0], point[1]),
    segmentIndex,
  };
}

export function closestPointOnPolyline(
  p: [number, number],
  coords: number[][],
  options?: { minFromEndpointKm?: number },
): ClosestPointOnSegment | null {
  if (coords.length < 2) return null;
  const minFromEndpoint = options?.minFromEndpointKm ?? LINE_SPLIT_ENDPOINT_MIN_KM;

  let best: ClosestPointOnSegment | null = null;
  for (let i = 0; i < coords.length - 1; i++) {
    const a = coords[i] as [number, number];
    const b = coords[i + 1] as [number, number];
    const hit = closestPointOnSegment(p, a, b, i);
    if (best == null || hit.distanceKm < best.distanceKm) best = hit;
  }
  if (!best) return null;

  const start = coords[0] as [number, number];
  const end = coords[coords.length - 1] as [number, number];
  const atLineStart =
    best.segmentIndex === 0 &&
    (best.t <= 1e-6 || haversineKm(best.point[0], best.point[1], start[0], start[1]) < minFromEndpoint);
  const atLineEnd =
    best.segmentIndex === coords.length - 2 &&
    (best.t >= 1 - 1e-6 || haversineKm(best.point[0], best.point[1], end[0], end[1]) < minFromEndpoint);
  if (atLineStart || atLineEnd) return null;

  return best;
}

export type LineSplitCandidate = {
  line: InfraObject;
  segmentIndex: number;
  snapLon: number;
  snapLat: number;
  distanceKm: number;
};

export function findLineSplitAtPoint(
  click: [number, number],
  infraObjects: InfraObject[],
  toleranceKm = LINE_SPLIT_HIT_TOLERANCE_KM,
): LineSplitCandidate | null {
  let best: LineSplitCandidate | null = null;

  for (const obj of infraObjects) {
    if (!isLineSubtype(obj.subtype)) continue;
    const coords = getLineCoordinates(obj);
    if (!coords || coords.length < 2) continue;
    const hit = closestPointOnPolyline(click, coords);
    if (!hit || hit.distanceKm > toleranceKm) continue;
    const candidate: LineSplitCandidate = {
      line: obj,
      segmentIndex: hit.segmentIndex,
      snapLon: roundCoord(hit.point[0]),
      snapLat: roundCoord(hit.point[1]),
      distanceKm: hit.distanceKm,
    };
    if (!best || candidate.distanceKm < best.distanceKm) best = candidate;
  }

  return best;
}

export function splitLineCoordinatesAt(
  coords: number[][],
  segmentIndex: number,
  split: [number, number],
): [number[][], number[][]] | null {
  if (coords.length < 2) return null;
  if (segmentIndex < 0 || segmentIndex >= coords.length - 1) return null;

  const q: [number, number] = [roundCoord(split[0]), roundCoord(split[1])];
  const first: number[][] = [...coords.slice(0, segmentIndex + 1), q];
  const second: number[][] = [q, ...coords.slice(segmentIndex + 1)];

  if (first.length < 2 || second.length < 2) return null;
  return [first, second];
}

export function infraLineGeometryPayloadFromCoords(
  coords: number[][],
  template: InfraObject,
  name: string,
): Pick<
  InfraObjectCreate,
  'name' | 'subtype' | 'layer_id' | 'lon' | 'lat' | 'end_lon' | 'end_lat' | 'coordinates'
> {
  const rounded = coords.map(([lo, la]) => [roundCoord(lo), roundCoord(la)] as [number, number]);
  return {
    name,
    subtype: template.subtype,
    layer_id: template.layer_id,
    lon: rounded[0]![0],
    lat: rounded[0]![1],
    end_lon: rounded[rounded.length - 1]![0],
    end_lat: rounded[rounded.length - 1]![1],
    coordinates: rounded,
  };
}

export function infraLinePayloadFromCoords(
  coords: number[][],
  template: InfraObject,
  name: string,
): InfraObjectCreate {
  const base = infraLineGeometryPayloadFromCoords(coords, template, name);
  if (!template.properties) return base;
  const props = { ...template.properties };
  delete props.coordinates;
  return { ...base, properties: props };
}

export function buildLineSplitPlan(
  line: InfraObject,
  segmentIndex: number,
  splitLon: number,
  splitLat: number,
  secondLineName: string,
): {
  firstPayload: Pick<
    InfraObjectCreate,
    'name' | 'subtype' | 'layer_id' | 'lon' | 'lat' | 'end_lon' | 'end_lat' | 'coordinates'
  >;
  secondPayload: InfraObjectCreate;
} | null {
  const coords = getLineCoordinates(line);
  if (!coords) return null;
  const parts = splitLineCoordinatesAt(coords, segmentIndex, [splitLon, splitLat]);
  if (!parts) return null;
  const [firstCoords, secondCoords] = parts;
  return {
    firstPayload: infraLineGeometryPayloadFromCoords(firstCoords, line, line.name),
    secondPayload: infraLinePayloadFromCoords(secondCoords, line, secondLineName),
  };
}

File: decision-matrix/backend/app/services/line_split.py
"""Split infrastructure line objects at a point (server-side, mirrors frontend lineSplit.ts)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from uuid import UUID

from app.models import InfrastructureObject
from app.services.spatial import line_coords_from_object
from app.services.road_graph import LINE_SPLIT_ENDPOINT_MIN_KM, haversine_km

LINE_SPLIT_HIT_TOLERANCE_KM = 0.3


@dataclass
class LineSplitPlan:
    first_coords: list[list[float]]
    second_coords: list[list[float]]
    second_name: str


def split_line_coordinates_at(
    coords: list[tuple[float, float]],
    segment_index: int,
    split_lon: float,
    split_lat: float,
) -> tuple[list[list[float]], list[list[float]]] | None:
    if len(coords) < 2:
        return None
    if segment_index < 0 or segment_index >= len(coords) - 1:
        return None
    q = [split_lon, split_lat]
    first = [[c[0], c[1]] for c in coords[: segment_index + 1]] + [q]
    second = [q] + [[c[0], c[1]] for c in coords[segment_index + 1 :]]
    if len(first) < 2 or len(second) < 2:
        return None
    return first, second


def build_line_split_plan(
    line: InfrastructureObject,
    segment_index: int,
    split_lon: float,
    split_lat: float,
    second_line_name: str,
) -> LineSplitPlan | None:
    coords = line_coords_from_object(line)
    parts = split_line_coordinates_at(coords, segment_index, split_lon, split_lat)
    if parts is None:
        return None
    first, second = parts
    return LineSplitPlan(first_coords=first, second_coords=second, second_name=second_line_name)


def line_geometry_fields_from_coords(
    coords: list[list[float]],
    *,
    name: str,
    subtype: str,
    layer_id: UUID,
    properties: dict[str, Any] | None = None,
) -> dict[str, Any]:
    props = dict(properties or {})
    props["coordinates"] = coords
    return {
        "name": name,
        "subtype": subtype,
        "layer_id": layer_id,
        "lon": coords[0][0],
        "lat": coords[0][1],
        "end_lon": coords[-1][0],
        "end_lat": coords[-1][1],
        "properties": props,
    }


def _segments_intersect(
    a1: tuple[float, float],
    a2: tuple[float, float],
    b1: tuple[float, float],
    b2: tuple[float, float],
) -> tuple[float, float] | None:
    """Planar segment intersection in lon/lat."""
    x1, y1 = a1
    x2, y2 = a2
    x3, y3 = b1
    x4, y4 = b2
    denom = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4)
    if abs(denom) < 1e-15:
        return None
    t = ((x1 - x3) * (y3 - y4) - (y1 - y3) * (x3 - x4)) / denom
    u = -((x1 - x2) * (y1 - y3) - (y1 - y2) * (x1 - x3)) / denom
    eps = 1e-9
    if eps < t < 1 - eps and eps < u < 1 - eps:
        ix = x1 + t * (x2 - x1)
        iy = y1 + t * (y2 - y1)
        return ix, iy
    return None


def find_segment_intersections(
    seg_a: tuple[tuple[float, float], tuple[float, float]],
    seg_b: tuple[tuple[float, float], tuple[float, float]],
) -> tuple[float, float] | None:
    return _segments_intersect(seg_a[0], seg_a[1], seg_b[0], seg_b[1])


def is_near_line_endpoint(
    lon: float,
    lat: float,
    coords: list[tuple[float, float]],
    *,
    min_km: float = LINE_SPLIT_ENDPOINT_MIN_KM,
) -> bool:
    if len(coords) < 2:
        return True
    start, end = coords[0], coords[-1]
    return (
        haversine_km(lon, lat, start[0], start[1]) < min_km
        or haversine_km(lon, lat, end[0], end[1]) < min_km
    )
